src/main/java/com/google/api/generator/engine/ast/AssignmentExpr.java

```diff
@@ -60,10 +60,12 @@ public AssignmentExpr build() {
                   "LHS type %s must match RHS type %s", lhsType.toString(), rhsType.toString()));
         }
       } else {
-        if (rhsType != TypeNode.NULL && !lhsType.isSupertypeOrEquals(rhsType)) {
+        if (!lhsType.equals(rhsType)
+            && rhsType != TypeNode.NULL
+            && !lhsType.isSupertypeOrEquals(rhsType)) {
           throw new TypeMismatchException(
               String.format(
-                  "LHS type %s must be a supertype of the RHS type %s",
+                  "LHS type %s must be a supertype of the RHS type %s, else they have to be boxed and primitive type equals.",
```

Comment:
Let's keep the boxed/primitive check separate, and add an `else if` for `rhsType.isPrimitiveType()`.

Reply:
Done.